File: src/main/java/com/epri/dss/delivery/impl/PDElementImpl.java
package com.epri.dss.delivery.impl;

import com.epri.dss.shared.impl.Complex;

import com.epri.dss.common.DSSClass;
import com.epri.dss.common.SolutionObj;
import com.epri.dss.common.impl.DSSCktElement;
import com.epri.dss.common.impl.DSSClassDefs;
import com.epri.dss.common.impl.DSSGlobals;
import com.epri.dss.delivery.PDElement;
import com.epri.dss.meter.MeterElement;

public class PDElementImpl extends DSSCktElement implements PDElement {

	protected double normAmps,
		emergAmps,
		faultRate,
		pctPerm,
		hrsToRepair;
	private int fromTerminal,
		toTerminal;  // set by meter zone for radial feeder
	protected boolean isShunt;

	private int numCustomers;
	private int totalCustomers;

	private PDElement parentPDElement;

	private MeterElement meterObj,   // upline energymeter
		sensorObj;  // upline sensor for this element for allocation and estimation

	private double overloadUE,
		overloadEEN;  // indicate amount of branch overload

	public PDElementImpl(DSSClass parClass) {
		super(parClass);

		isShunt = false;
		fromTerminal = 1;
		numCustomers = 0;
		totalCustomers = 0;
		sensorObj = null;
		meterObj = null;
		parentPDElement = null;
		objType = DSSClassDefs.PD_ELEMENT;
	}

	@Override
	public void getCurrents(Complex[] curr) {
		int i;
		DSSGlobals globals = DSSGlobals.getInstance();

		try {
			if (isEnabled()) {
				SolutionObj sol = globals.getActiveCircuit().getSolution();
				for (i = 0; i < YOrder; i++)
					VTerminal[i] = sol.getNodeV()[nodeRef[i]];

				YPrim.MVMult(curr, VTerminal);
			} else {
				for (i = 0; i < YOrder; i++)
					curr[i] = Complex.ZERO;
			}
		} catch (Exception e) {
			globals.doErrorMsg(("Trying to Get Currents for Element: " + getName() + "."), e.getMessage(),
					"Has circuit been solved?", 660);
		}
	}

	public Complex getExcessKVANorm (int idxTerm) {
		Complex result;

		if ((normAmps == 0.0) || !isEnabled()) {
			overloadEEN = 0.0;
			return Complex.ZERO;
		}

		Complex kVA = getPower(idxTerm).multiply(0.001);  // also forces computation of current into iTemp
		double Factor = maxTerminalOneIMag() / normAmps - 1.0;
		if (Factor > 0.0) {
			overloadEEN = Factor;
			Factor = 1.0 - 1.0 / (Factor + 1.0);  // to get factor
			result = kVA.multiply(Factor) ;
		} else {
			overloadEEN = 0.0;
			result = Complex.ZERO;
		}
		/* **********DEBUG CODE: Use DLL Debug file  *** */
		/* ****    WriteDLLDebugFile(String.format("%s.%s: Terminal=%u Factor=%.7g kW=%.7g kvar=%.7g Normamps=%.7g Overload_EEN=%.7g Result=%.7g +j %.7g ",
			ParentClass.getName(), getName(), ActiveTerminalIdx, Factor, kVA.re, kVA.im, NormAmps, Overload_EEN, Result.re, Result.im));
		*/

		return result;
	}

	public Complex getExcessKVAEmerg(int idxTerm) {
		Complex result;

		if ((getEmergAmps() == 0.0) || !isEnabled()) {
			overloadUE = 0.0;
			return Complex.ZERO;
		}

		Complex kVA = getPower(idxTerm).multiply(0.001);  // also forces computation of current into iTemp

		double Factor = maxTerminalOneIMag() / getEmergAmps() - 1.0;
		if (Factor > 0.0) {
			overloadUE = Factor;
			Factor = 1.0 - 1.0 / (Factor + 1.0);  // to get excess
			result = kVA.multiply(Factor);
		} else {
			overloadUE = 0.0;
			result = Complex.ZERO;
		}

		return result;
	}

	@Override
	public void initPropertyValues(int arrayOffset) {

		propertyValue[arrayOffset + 1] = "400";  // normAmps   TODO Check zero based indexing
		propertyValue[arrayOffset + 2] = "600";  // emerAamps
		propertyValue[arrayOffset + 3] = "0.1";  // faultRate
		propertyValue[arrayOffset + 4] = "20";   // pctPerm
		propertyValue[arrayOffset + 5] = "3";    // hrsToRepair

		super.initPropertyValues(arrayOffset + 5);
	}

	public double getNormAmps() {
		return normAmps;
	}

	public void setNormAmps(double amps) {
		normAmps = amps;
	}

	public double getEmergAmps() {
		return emergAmps;
	}

	public void setEmergAmps(double amps) {
		emergAmps = amps;
	}

	public double getFaultRate() {
		return faultRate;
	}

	public void setFaultRate(double rate) {
		faultRate = rate;
	}

	public double getPctPerm() {
		return pctPerm;
	}

	public void setPctPerm(double pct) {
		pctPerm = pct;
	}

	public double getHrsToRepair() {
		return hrsToRepair;
	}

	public void setHrsToRepair(double hrs) {
		hrsToRepair = hrs;
	}

	public int getFromTerminal() {
		return fromTerminal;
	}

	public void setFromTerminal(int terminal) {
		fromTerminal = terminal;
	}

	public int getToTerminal() {
		return toTerminal;
	}

	public void setToTerminal(int terminal) {
		toTerminal = terminal;
	}

	public boolean isShunt() {
		return isShunt;
	}

	public void setShunt(boolean value) {
		isShunt = value;
	}

	public int getNumCustomers() {
		return numCustomers;
	}

	public void setNumCustomers(int num) {
		numCustomers = num;
	}

	public int getTotalCustomers() {
		return totalCustomers;
	}

	public void setTotalCustomers(int total) {
		totalCustomers = total;
	}

	public PDElement getParentPDElement() {
		return parentPDElement;
	}

	public void setParentPDElement(PDElement parent) {
		parentPDElement = parent;
	}

	public MeterElement getMeterObj() {
		return meterObj;
	}

	public void setMeterObj(MeterElement meter) {
		meterObj = meter;
	}

	public MeterElement getSensorObj() {
		return sensorObj;
	}

	public void setSensorObj(MeterElement sensor) {
		sensorObj = sensor;
	}

	public double getOverloadUE() {
		return overloadUE;
	}

	public void setOverload_UE(double overload) {
		overloadUE = overload;
	}

	public double getOverloadEEN() {
		return overloadEEN;
	}

	public void setOverloadEEN(double overload) {
		overloadEEN = overload;
	}

}

File: pydss/delivery/PDElementImpl.py
from dss.common.impl.DSSGlobals import (DSSGlobals,)
from dss.common.impl.DSSClassDefs import (DSSClassDefs,)
from dss.delivery.PDElement import (PDElement,)
from dss.common.impl.DSSCktElement import (DSSCktElement,)
# from org.apache.commons.math.complex.Complex import (Complex,)


class PDElementImpl(DSSCktElement, PDElement):
    normAmps = None
    emergAmps = None
    faultRate = None
    pctPerm = None
    hrsToRepair = None
    fromTerminal = None
    toTerminal = None
    # set by meter zone for radial feeder
    isShunt = None
    numCustomers = None
    totalCustomers = None
    parentPDElement = None
    meterObj = None
    sensorObj = None
    # upline energymeter
    # upline sensor for this element for allocation and estimation
    overloadUE = None
    overloadEEN = None
    # indicate amount of branch overload

    def __init__(self, parClass):
        super(PDElementImpl, self)(parClass)
        self.isShunt = False
        self.fromTerminal = 1
        self.numCustomers = 0
        self.totalCustomers = 0
        self.sensorObj = None
        self.meterObj = None
        self.parentPDElement = None
        self.objType = DSSClassDefs.PD_ELEMENT

    def getCurrents(self, curr):
        try:
            if self.isEnabled():
                sol = DSSGlobals.activeCircuit.getSolution()
                _0 = True
                i = 0
                while True:
                    if _0 is True:
                        _0 = False
                    else:
                        i += 1
                    if not (i < self.YOrder):
                        break
                    self.VTerminal[i] = sol.getNodeV()[self.nodeRef[i]]
                self.YPrim.vMult(curr, self.VTerminal)
            else:
                _1 = True
                i = 0
                while True:
                    if _1 is True:
                        _1 = False
                    else:
                        i += 1
                    if not (i < self.YOrder):
                        break
                    curr[i] = Complex.ZERO
        except Exception, e:
            DSSGlobals.doErrorMsg('Trying to Get Currents for Element: ' + self.getName() + '.', e.getMessage(), 'Has circuit been solved?', 660)

    def getExcessKVANorm(self, idxTerm):
        if (self.normAmps == 0.0) or (not self.isEnabled()):
            self.overloadEEN = 0.0
            return Complex.ZERO
        kVA = self.getPower(idxTerm).multiply(0.001)
        # also forces computation of current into iTemp
        Factor = (self.maxTerminalOneIMag() / self.normAmps) - 1.0
        if Factor > 0.0:
            self.overloadEEN = Factor
            Factor = 1.0 - (1.0 / (Factor + 1.0))
            # to get factor
            result = kVA.multiply(Factor)
        else:
            self.overloadEEN = 0.0
            result = Complex.ZERO
        # **********DEBUG CODE: Use DLL Debug file  ***
        # ****    WriteDLLDebugFile(String.format("%s.%s: Terminal=%u Factor=%.7g kW=%.7g kvar=%.7g Normamps=%.7g Overload_EEN=%.7g Result=%.7g +j %.7g ",
        # 			ParentClass.getName(), getName(), ActiveTerminalIdx, Factor, kVA.re, kVA.im, NormAmps, Overload_EEN, Result.re, Result.im));

        return result

    def getExcessKVAEmerg(self, idxTerm):
        if (self.getEmergAmps() == 0.0) or (not self.isEnabled()):
            self.overloadUE = 0.0
            return Complex.ZERO
        kVA = self.getPower(idxTerm).multiply(0.001)
        # also forces computation of current into iTemp
        Factor = (self.maxTerminalOneIMag() / self.getEmergAmps()) - 1.0
        if Factor > 0.0:
            self.overloadUE = Factor
            Factor = 1.0 - (1.0 / (Factor + 1.0))
            # to get excess
            result = kVA.multiply(Factor)
        else:
            self.overloadUE = 0.0
            result = Complex.ZERO
        return result

    def initPropertyValues(self, arrayOffset):
        self.propertyValue[arrayOffset + 1] = '400'
        # normAmps   TODO Check zero based indexing
        self.propertyValue[arrayOffset + 2] = '600'
        # emerAamps
        self.propertyValue[arrayOffset + 3] = '0.1'
        # faultRate
        self.propertyValue[arrayOffset + 4] = '20'
        # pctPerm
        self.propertyValue[arrayOffset + 5] = '3'
        # hrsToRepair
        super(PDElementImpl, self).initPropertyValues(arrayOffset + 5)

    def getNormAmps(self):
        return self.normAmps

    def setNormAmps(self, amps):
        self.normAmps = amps

    def getEmergAmps(self):
        return self.emergAmps

    def setEmergAmps(self, amps):
        self.emergAmps = amps

    def getFaultRate(self):
        return self.faultRate

    def setFaultRate(self, rate):
        self.faultRate = rate

    def getPctPerm(self):
        return self.pctPerm

    def setPctPerm(self, pct):
        self.pctPerm = pct

    def getHrsToRepair(self):
        return self.hrsToRepair

    def setHrsToRepair(self, hrs):
        self.hrsToRepair = hrs

    def getFromTerminal(self):
        return self.fromTerminal

    def setFromTerminal(self, terminal):
        self.fromTerminal = terminal

    def getToTerminal(self):
        return self.toTerminal

    def setToTerminal(self, terminal):
        self.toTerminal = terminal

    def isShunt(self):
        return self.isShunt

    def setShunt(self, value):
        self.isShunt = value

    def getNumCustomers(self):
        return self.numCustomers

    def setNumCustomers(self, num):
        self.numCustomers = num

    def getTotalCustomers(self):
        return self.totalCustomers

    def setTotalCustomers(self, total):
        self.totalCustomers = total

    def getParentPDElement(self):
        return self.parentPDElement

    def setParentPDElement(self, parent):
        self.parentPDElement = parent

    def getMeterObj(self):
        return self.meterObj

    def setMeterObj(self, meter):
        self.meterObj = meter

    def getSensorObj(self):
        return self.sensorObj

    def setSensorObj(self, sensor):
        self.sensorObj = sensor

    def getOverloadUE(self):
        return self.overloadUE

    def setOverload_UE(self, overload):
        self.overloadUE = overload

    def getOverloadEEN(self):
        return self.overloadEEN

    def setOverloadEEN(self, overload):
        self.overloadEEN = overload
